branchout

```diff
@@ -86,6 +86,71 @@ function branchout_os() {
   echo "${OS}"
 }
 
+# Update all Git URLs in 3 levels .git/config files that match the branchout Git base URL by prefix
+function branchout_relocate() {
+  if [ $# -ne 1 ]; then
+    fail "branchout relocate requires exactly one argument: The new Git URL base to apply to all remotes in all repos that currently use ${BRANCHOUT_GIT_BASEURL}"
+  fi
+
+  NEW_GIT_BASEURL="${1}"
+
+  if [ "${BRANCHOUT_GIT_BASEURL}" == "${NEW_GIT_BASEURL}" ]; then
+    fail "Old prefix and new prefix are the same, no point in continuing: ${BRANCHOUT_GIT_BASEURL}"
+  fi
+
+  echo "Relocating all Git repos from ${BRANCHOUT_GIT_BASEURL} to ${NEW_GIT_BASEURL}"
+  echo "in ${PROJECTION_DIRECTORY} and all nested repos 1 and 2 levels deep."
+
+  # Start with the entire git URL
+  OUR_FULL_URL="$(git --git-dir "${PROJECTION_DIRECTORY}/.git" remote get-url origin 2>/dev/null)"
+  echo "This is the current Git URL for ${PROJECTION_DIRECTORY}: ${OUR_FULL_URL}"
+
+  # Build our new URL
+  NEW_FULL_URL="${OUR_FULL_URL//${BRANCHOUT_GIT_BASEURL}/${NEW_GIT_BASEURL}}"
+  echo "This is the new Git URL for ${PROJECTION_DIRECTORY}: ${NEW_FULL_URL}"
+
+  echo "Validating the Git repo pointed at by the new URL above works..."
+  if git ls-remote "${NEW_FULL_URL}" > /dev/null; then
+    echo "New URL exists and you have access, proceeding..."
+  else
+    fail "New URL is bad, repo does not exist, or you have no access to it: ${NEW_FULL_URL}"
+  fi
+
+  # Collect together all of the Git config files to process
+  CONFIG_FILES=("${PROJECTION_DIRECTORY}/.git/config")
+
+  GROUP_CONFIG_FILE_PATTERN="${PROJECTION_DIRECTORY}/*/.git/config"
+  if compgen -G "${GROUP_CONFIG_FILE_PATTERN}" > /dev/null; then
+    GROUP_CONFIG_FILES=()
+    while IFS='' read -r configFile; do GROUP_CONFIG_FILES+=("$configFile"); done < <(compgen -G "${GROUP_CONFIG_FILE_PATTERN}")
+    CONFIG_FILES=("${CONFIG_FILES[@]}" "${GROUP_CONFIG_FILES[@]}")
+    echo "Found ${#GROUP_CONFIG_FILES[@]} group .git/config files to process"
+  else
+    echo "No group .git/config files found to process."
+  fi
+
+  PROJECT_CONFIG_FILE_PATTERN="${PROJECTION_DIRECTORY}/*/*/.git/config"
+  if compgen -G "${PROJECT_CONFIG_FILE_PATTERN}" > /dev/null; then
+    PROJECT_CONFIG_FILES=()
+    while IFS='' read -r configFile; do PROJECT_CONFIG_FILES+=("$configFile"); done < <(compgen -G "${PROJECT_CONFIG_FILE_PATTERN}")
+    CONFIG_FILES=("${CONFIG_FILES[@]}" "${PROJECT_CONFIG_FILES[@]}")
+    echo "Found ${#PROJECT_CONFIG_FILES[@]} project .git/config files to process"
+  else
+    echo "No project .git/config files found to process."
+  fi
```

Comment:
The initial impl wasn't using the loop, but given you have it now why continue with the glob, why not just pipe find into the loop and check for config?

CONFIG_FILES=$(find . -maxdepth 3 -type d -name '.git' | while read -r gitDirectory; do test -f ${gitDirectory}/config && echo  ${gitDirectory}/config; done)

Reply:
Fair enough. Give me a few to rejig it

Reply:
No, find . is no good, but I know what is.

Reply:
SC2031 has you by the short and curlies. :-D